development/workflow.rst

```diff
@@ -0,0 +1,305 @@
+
+The master branch
+-----------------
+
+``master`` is the main integration branch for our repositories.
+Development is not done directly on the ``master`` branch, but instead on *ticket branches*.
+*Trivial* documentation fixes are the only scenarios where working directly on ``master`` and by-passing the code review process is permitted.
+
+The Git history of ``master`` should never be re-written with force pushes.
+
+User branches
+-------------
+
+You can do experimental, proof-of-concept work in 'user branches.'
+
+These branches are named
+
+.. code-block:: text
+
+   u/{{username}}/{{topic}}
+
+User branches can be pushed to GitHub to enable collaboration and communication.
+Colleagues should be cognizant that user branches are for prototyping before offering unsolicited code reviews.
+
+Developers can feel free to rebase and force push work to their personal user branches.
+
+A user branch *cannot* be merged into master; it must be converted into a *ticket branch* first.
+
+Ticket branches
+---------------
+
+Ticket branches are associated with a JIRA ticket.
+Only ticket branches can be merged into ``master``.
+
+If the JIRA ticket is named ``DM-NNNN``, then the ticket branch will be named
+
+.. code-block:: text
+
+   tickets/DM-NNNN
+
+A ticket branch can be made by branching off an existing user branch.
+This is a great way to formalize experimental work and shaping it into an LSST software contribution.
+
+When code on a ticket branch is ready for review and merging, follow the :ref:`code review process documentation <git-code-review>`.
+
+.. _git-code-review:
+
+DM Code Review and Merging Process
+==================================
+
+When development on your ticket branch is complete, there is a standard process for reviewing your work and merging it into the ``master`` integration branch.
+
+Review preparation
+------------------
+
+Before submitting your work, we recommend that you organize commits, improve commit messages, and ensure that your work is made against the latest commits on ``master`` with an `interactive rebase <https://help.github.com/articles/about-git-rebase/>`_.
+A common pattern is:
+
+.. code-block:: bash
+
+   git checkout master
+   git pull
+   git checkout tickets/DM-NNNN
+   git rebase -i master
+   # interactive rebase
+   git push --force
+
+Use `Jenkins at ci.lsst.codes <https://ci.lsst.codes/job/stack-os-matrix/build?delay=0sec>`_ to run the Stack's tests with your ticket branch work.
+With Jenkins you can specify the name (or names) of the ticket branches in repositories you are developing; Jenkins will test your work against the ``master`` branch of other Stack repositories by default.
+
+You can monitor builds in the “Bot: Jenkins” HipChat room.
+
+Assign a reviewer
+-----------------
+
+On GitHub, `create a pull request <https://help.github.com/articles/creating-a-pull-request/>`_ for your ticket branch.
+
+.. FIXME policy for naming pull requests?
```

Comment:
PLEASE include more context than just the ticket number in the pull request title.  There are a lot of ticket numbers flying by at any one time, and those of us who haven't memorised them all need to have that context in the subject line of the e-mail notifications that github sends so we know if we should look closer.  I suggest putting the ticket number and the Jira title of the ticket.